src/libc/stdio/putc.3
.\"	$NetBSD: putc.3,v 1.10 2003/04/16 13:34:44 wiz Exp $
.\"
.\" Copyright (c) 1990, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Chris Torek and the American National Standards Committee X3,
.\" on Information Processing Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)putc.3	8.1 (Berkeley) 6/4/93
.\"
.Dd April 25, 2001
.Dt PUTC 3
.Os
.Sh NAME
.Nm fputc ,
.Nm putc ,
.Nm putchar ,
.Nm putc_unlocked ,
.Nm putchar_unlocked ,
.Nm putw
.Nd output a character or word to a stream
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In stdio.h
.Ft int
.Fn fputc "int c" "FILE *stream"
.Ft int
.Fn putc "int c" "FILE *stream"
.Ft int
.Fn putchar "int c"
.Ft int
.Fn putc_unlocked "int c" "FILE *stream"
.Ft int
.Fn putchar_unlocked "int c"
.Ft int
.Fn putw "int w" "FILE *stream"
.Sh DESCRIPTION
The
.Fn fputc
function
writes the character
.Fa c
(converted to an ``unsigned char'')
to the output stream pointed to by
.Fa stream .
.Pp
.Fn putc
acts essentially identically to
.Fn fputc ,
but is a macro that expands in-line.
It may evaluate
.Fa stream
more than once, so arguments given to
.Fn putc
should not be expressions with potential side effects.
.Pp
.Fn putchar
is identical to
.Fn putc
with an output stream of
.Em stdout .
.Pp
The
.Fn putc_unlocked
and
.Fn putchar_unlocked
functions provide functionality identical to that of
.Fn putc
and
.Fn putchar ,
respectively, but do not perform implicit locking of the streams they
operate on.
In multi-threaded programs they may be used
.Em only
within a scope in which the stream
has been successfully locked by the calling thread using either
.Xr flockfile 3
or
.Xr ftrylockfile 3 ,
and may later be released using
.Xr funlockfile 3 .
.Pp
The
.Fn putw
function
writes the specified
.Em int
to the named output
.Fa stream .
.Sh RETURN VALUES
The functions,
.Fn fputc ,
.Fn putc
and
.Fn putchar
return the character written.
If an error occurs, the value
.Dv EOF
is returned.
The
.Fn putw
function
returns 0 on success;
.Dv EOF
is returned if
a write error occurs,
or if an attempt is made to write a read-only stream.
.Sh SEE ALSO
.Xr ferror 3 ,
.Xr fopen 3 ,
.Xr getc 3 ,
.Xr stdio 3
.Sh STANDARDS
The functions
.Fn fputc ,
.Fn putc ,
and
.Fn putchar ,
conform to
.St -ansiC .
The functions
.Fn putc_unlocked
and
.Fn putchar_unlocked
conform to
.St -p1003.1-96 .
A function
.Fn putw
function appeared in
.At v6 .
.Sh BUGS
The size and byte order of an
.Em int
varies from one machine to another, and
.Fn putw
is not recommended for portable applications.


RELEASE-NOTES
Release notes for BitKeeper version 4.1 (released 12-Oct-2007)

Major features
    BAM support.  BAM stands for "Binary Asset Management" and it adds 
    support to BK for versioning large binaries.  It solves two problems:
    a) one or more binary files that are frequently changed.
    b) collections of many large binaries where you only need a subset.
    The way it solves this is to introduce the concept of BAM server[s].
    A BAM server manages a collection of binaries for one or more BAM
    clients.  BAM clients may have no data present; when it is needed
    the data is fetched from the BAM server.  
    
    In the first case above, only the tip will be fetched.  Imagine that
    you have 100 deltas, each 10MB in size.  The history is 1GB but you
    only need 10MB in your clone.

    In the second case, imagine that you have thousands of game assets
    distributed across multiple directories.  You typically work only
    in one directory at a time.  You will only need to fetch the subset
    of files that you need, the rest of the repository will have the
    history of what changed but no data (so bk log will work but 
    bk cat will have to go fetch the data).

    BAM files are currently limited to 32 bits in size.

    See bk help Howto-BAM and bk help bam for more information.

    -----

    bk repair makes use of the new -@<url> and BAM infrastructure to run
    much more quickly.  Our stress tests for this amount to a loop, cloning
    each tagged version of BitKeeper, removing each top level directory,
    and running bk repair to restore.  Works fine, and works quickly.
    Please let us know if you find problems with this feature, for now it
    is enabled when you run check with -ff (2 "f" chars) but in the future
    we intend to turn it on in any repo with autofix configured.
    See bk help repair for more information.

Summary of changes

Performance fixes
    Repositories with large numbers of changesets will run faster when
    doing pull/push/changes.  For example, in the Linux kernel 2.6 tree
    with 128 thousand changesets, the time to do "bk changes -r+" is 1/2
    of what it was in bk-4.0.2.  Null pulls/pushes should see similar
    performance gains.

    Pulls which update repositories with large numbers of deleted files
    will run significantly faster.

Windows Vista support
    The Windows version of BitKeeper now runs on Windows Vista on 32-bit
    architectures.  64-bit support for Vista and/or XP will be in a future
    release.

Triggers
    This release introduces the concept of a trigger path.  The trigger
    path may be set in the configuration file to create a set of directories
    to search.  All triggers found in all directories are run (unless one
    is a pre- trigger and it fails).
    See bk help config-etc.

    This release introduces support for pre-undo triggers.  
    See bk help triggers.

Checkout mode
    A new checkout mode, checkout:last, has been added.  For clones, it
    is the same as checkout:none, but after the file has been checked out
    (read-only or read-write), BK will maintain that state.  The main
    use is for repositories with large numbers of large BAM binaries.
    The mode makes it possible to work on a subset of those files which
    gives you the effect of a sparse tree.  If said repository was in
    checkout:get or checkout:edit mode then the tips of all files will
    be present, i.e., the repository will not be sparse.

BK/Web
    BK/Web now uses MD5 keys for all URLs and revision numbers.  This
    allows bookmarks to continue to work even after revision numbers are
    changed by merges.  For this reason, the "Link to this page" links have
    been removed.
    Note the URL syntax has not changed from the 4.0 release, so old URLs
    will still work correctly.
    
    BK/Web now lists extra files while browsing the source tree; they are
    at the end of the directory listing which is { dirs, BK files, extras }.
    Extra files are only listed if they are not ignored.  Extra files 
    may be clicked on and viewed.

    BK/Web has links on the file names in the history/annotate pages that
    will let you download a copy of that file with no markups.

SSH URLs
    BitKeeper now allows an SSH URL to specify a non-standard port,
    for example ssh://host.example.com:5555/repopath.  See bk help url.

Remote command execution
    The command syntax 'bk -@URL command' allows a bk command to be
    run on a remote repository with the output returned locally.
    See 'bk help bk' for more information.
    Example showing work in progress (diffs and new files) in a remote repo:

    	bk -@bk://work/BAM -r diffs -Nur

Configuration
    nosync: on

    By default, BK does a sync in the resolver after updating your repository.
    The reason it does not fsync() individual files is that on some Linux
    file systems, the fsync() is like a sync() and sync() is global.  Doing
    it over and over again hurt server performance.  So we do one sync after
    applying all the files.

    Some customers prefer performance over safety; they can set this variable
    and get no syncs at all.  We don't recommend that.

bk admin
    The -H option formerly checked sfile structure, limited to AT&T
    SCCS features, and checked whether the file contents were ASCII.
    As BitKeeper now fully supports non-ASCII characters, this option
    is deprecated.  Use -h, -hh, or -hhh instead.  See bk help admin.

bk bkd
    The -p option now allows you to restrict allowed connections to
    those originating from a specified IP address.  A common usage is
    "-p127.0.0.1:", which accepts connections only on the default port
    (14690) on localhost.  See bk help bkd.

bk changes
    The search facility ("-/<pattern>/") now supports two additional
    suffixes:  "g", which indicates that <pattern> is a glob, not a regular
    expression; and "t", which restricts the search to tag names.  See bk
    help changes.


src/libc/stdio/putc.c
/*	$NetBSD: putc.c,v 1.10 2003/07/18 21:52:17 nathanw Exp $	*/

/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Chris Torek.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
#if 0
static char sccsid[] = "@(#)putc.c	8.1 (Berkeley) 6/4/93";
#else
__RCSID("$NetBSD: putc.c,v 1.10 2003/07/18 21:52:17 nathanw Exp $");
#endif
#endif /* LIBC_SCCS and not lint */

#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include "reentrant.h"
#include "local.h"

/*
 * A subroutine version of the macro putc.
 */
#undef putc
#undef putc_unlocked

int
putc(c, fp)
	int c;
	FILE *fp;
{
	int r;

	assert(fp != NULL);

	FLOCKFILE(fp);
	r = __sputc(c, fp);
	FUNLOCKFILE(fp);
	return r;
}

int
putc_unlocked(c, fp)
	int c;
	FILE *fp;
{

	assert(fp != NULL);

	return (__sputc(c, fp));
}
